lib/feedsme/index.js

```diff
@@ -54,153 +64,223 @@ function pkgName(pkg) {
-Feedsme.prototype.trigger = function trigger(env, data, done) {
-  var feedsme = this;
-  var app = feedsme.app;
-  var BuildHead = app.models.BuildHead;
-  var Dependent = app.models.Dependent;
-  var Package = app.models.Package;
-  var Version = app.models.Version;
-  var npm = app.config.get('npm');
-  var name = data.name;
-
-  app.log.debug(`Trying to rebuild the dependents of ${ name }`);
-
-  Dependent.get(name, function found(error, details) {
-    if (error || !details) {
-      if (error) {
-        app.log.debug(`Failed to retrieve dependents of ${ name }: ${ error.message }`, {
-          innerErrors: error.innerErrors,
-          stack: error.stack
-        });
-      } else {
-        app.log.debug(`Found no dependents ${ name }`);
-      }
+Feedsme.prototype.triggerDep = async function triggerDep({ env, dep, releaseLine, pkg }) {
+  const { name, version } = pkg;
+  const spec = {
+    previousVersion: releaseLine.previousVersion,
+    pkg: name,
+    version
+  };
+  const publish = env === 'dev';
+
+  // If we are a publish, create a release-line
+  // NOTE: this version can be innaccurate if the npm publish was done with
+  // a tag that !== latest. We should handle this case at some point, fetch
+  // package for version or something
+  if (publish) {
+    this.log.info('create release-line', spec);
+    await this.release.create({
+      previousVersion: releaseLine.previousVersion,
+      pkg: name,
+      version
+    });
+  }
 
-      return done(error);
-    }
+  this.log.debug(`Trying to rebuild the dependents of ${ name }`);
+  if (!dep || (dep && dep.dependents.length === 0))
+    return this.log.debug(`Found no dependents for ${ name }`);
 
-    async.eachLimit(details.dependents, 5, function notify(pack, fn) {
-      app.log.debug(`Finding the package ${ pack }`);
+  const limiter = pLimit(this.conc);
 
-      Package.get(pack, function packages(err, pkg) {
-        if (err || !pkg) {
-          if (err) {
-            app.log.info(`Failed to look up package ${ pack } due to error ${ err.message }`, {
-              innerErrors: err.innerErrors,
-              stack: err.stack
-            });
-          } else {
-            app.log.debug(`Package ${ pack } not found in cassandra`);
-          }
+  await Promise.all(
+    dep.dependents.map(name => {
+      return limiter(() => this._triggerOne({ rootPkg: pkg, env, name, releaseLine }));
+    })
+  );
+};
 
-          return fn(err);
-        }
+Feedsme.prototype._triggerOne = async function triggerOne({ env, rootPkg, name, releaseLine }) {
+  const { Package, Version, BuildHead } = this.models;
+  //
+  // Grab the releaseVersion for the dependent if it exists in the
+  // releaseLine. This is used for correct promotion from DEV -> TEST and
+  // TEST -> PROD. Main package version owns the dependent package versions
+  // which get promoted that were established upon publish
+  //
+  const releaseVersion = releaseLine.dependents && releaseLine.dependents[name];
```

Comment:
are release lines back-ported in cassandra?  what happens here if there's an old build running through all this code and doesn't have a release line?

Reply:
A migration script would be independent of this but it's a good call-out.